rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,296 @@
+# RFC 120 object Service
+* Comments: [#116](https://api.github.com/repos/taskcluster/taskcluster-rfcs/issues/120)
+* Proposed by: @jhford
+
+# Summary
+
+Taskcluster tasks often produce artifacts during the course of being run.
+Examples include browser distribution archives, json metadata and log files.
+We currently store artifacts in a single region of S3 and split management of
+artifacts between the Queue and Cloud Mirror.  This project is to design an
+object storage api which drastically simplifies the management of objects from
+the Queue's perspective while at the same time handling the cross-region
+replication and cache management of Cloud Mirror as well as enabling
+cross-cloud support.
+
+The object service should not be directly exposed to end users as a Rest API.
+It will instead be a generalised API that the Queue uses to facilitate artifact
+creation from tasks.  In other words, the Queue gates operations to the object
+service.  The service will be exposed to endusers by means of a `Location`
+header value on a `300` series redirect or by relaying responses.
+
+The Queue will now store only the name of objects and would treat the name of
+an object as an opaque identifier and defer all management of the actual bytes
+to this system.  The Queue would also stop doing IP to object store mappings.
+
+This RFC is to decide on what the objects service is and its interface, not
```

Comment:
This RFC describes the interface of the Object Service; implementations may vary.